### ISSUE TITLE
Calculator: sign-change will sometimes decrease absolute value

### ISSUE THREAD

**Issue by NonStandardModel** [SerenityOS/serenity#7484]
Continuous pressing of change-sign button will decrease the value for some decimal numbers. Most extreme case I could find is `0.37`, which will decrease for 9 sign-changes down to `0.3699991`.

**Comment by BenWiederhake** [comment]
It appears that `Keypad.cpp` implemented its own version of float-to-string-formatting, and does so by effectively [rounding towards zero](https://github.com/SerenityOS/serenity/blob/master/Userland/Applications/Calculator/Keypad.cpp#L142).

I see at least three ways around this:
- Hack CalculatorWidget to toggle `m_negative` directly, probably also setting `keypad.m_state = State::External;` This is awful because it violates several programming principles.
- Implementing `Keypad::set_value` by doing a `String::formatted("{.6}", value)` and parsing it from there. It [kinda looks like](https://github.com/SerenityOS/serenity/blob/master/AK/Format.cpp#L358) it's trying to rounding in a less-incorrect way. This is awful because it is still not numerically stable.
- Introducing `double m_externally_set;` to `Keypad`, which is used if and only if `m_state == State::External`. This actually only requires additional work in `Keypad::to_string`, so maybe try this?

**Comment by MaxWipfli** [comment]
The real solution would probably be to limit the calculator to e.g. 9 decimals, and just store the value as a `BigInt` (with arbitrary length), but since we don't have that, this is not possible right now.

**Comment by linusg** [comment]
> but since we don't have that, this is not possible right now.

What's "that"? We do have arbitrary-length big integers :)

**Comment by BenWiederhake** [comment]
`Keypad` already uses it's own hand-rolled big int to store/edit the number. The conversion between this format and `double`, and the back-and-forth between the formats, that seems to be the issue. I don't really see how a new implementation would fix that; instead, I still recommend [these approaches](https://github.com/SerenityOS/serenity/issues/7484#issuecomment-850562802).

Or do you mean doing the operations on BigDecimals? In that case, good luck implementing exponentiation (including sqrt) and logarithms.

**Comment by MaxWipfli** [comment]
> What's "that"? We do have arbitrary-length big integers :)

We do? Okey, that's on me then! I probably just assumed this because of some stuff I read on Discord about `u128`, from which I wrongly implied we didn't have hat.

**Comment by MaxWipfli** [comment]
> `Keypad` already uses it's own hand-rolled big int to store/edit the number. The conversion between this format and `double`, and the back-and-forth between the formats, that seems to be the issue. I don't really see how a new implementation would fix that; instead, I still recommend [these approaches](https://github.com/SerenityOS/serenity/issues/7484#issuecomment-850562802).
> 
> Or do you mean doing the operations on BigDecimals? In that case, good luck implementing exponentiation (including sqrt) and logarithms.

Yeah, you're right. It probably makes sense to use the current format (or something similar) for input, as this avoids some rounding errors while inputting, but it could make sense to use the `m_externally_set` to avoid converting preexisting `double` values to the custom format and back to `double`.
